### ISSUE TITLE
Annotated generic variable loses its type argument when initialized with an unspecialized constructor

### ISSUE THREAD

**Issue by AnJunHao** [astral-sh/ty#3138]
### Summary

```python

from typing import cast

class Parent[T]: ...

def func[T](param: Parent[T]) -> T:
    return cast(T, ...)

p1: Parent[str] = cast(Parent[str], Parent())
result_p1 = func(p1)
reveal_type(result_p1) # Revealed type: `str` (revealed-type) [Ln 10, Col 13]

p2: Parent[str] = Parent()
result_p2 = func(p2)
reveal_type(result_p2) # Revealed type: `Unknown` (revealed-type) [Ln 14, Col 13]

```

A variable explicitly annotated as Parent[str] should be treated as Parent[str] afterward, so passing it to func should infer T = str and return str. The checker instead reveals Unknown when the variable is initialized with Parent(), even though the same annotated type works correctly when the value is cast to Parent[str]. This suggests the initializer’s unspecialized generic type is overriding the declared variable type during later inference.

See playground:

[https://play.ty.dev/70744710-58a9-431a-9abf-49fe11e883ac](url)

### Version

ty 0.0.25

**Comment by AlexWaygood** [comment]
Thanks! This is #136
